sklearn/utils/_sorting.pyx

```diff
@@ -1,22 +1,138 @@
 from cython cimport floating
+from libc.math cimport log as ln
 
-cdef inline void dual_swap(
-    floating* darr,
-    intp_t *iarr,
-    intp_t a,
-    intp_t b,
+from ._typedefs cimport intp_t
+
+# TODO: In order to support discrete distance metrics, we need to have a
+# simultaneous sort which breaks ties on indices when distances are identical.
+# The best might be using a std::stable_sort and a Comparator which might need
+# an Array of Structures (AoS) instead of the Structure of Arrays (SoA)
+# currently used. Alternatively, we can find a stable algorithm for SoA and
+# adapt it so that it is simultaneous.
+
+# Utilities functions
+
+# TODO: Factor code also present in `tree._utils` or use `libc.math.log2` directly
+cdef inline double log(double x) nogil:
+    return ln(x) / ln(2.0)
```

Comment:
I'd be in favor of using log2 directly, and especially not name it ``log`` which is confusing because it's usually used for the base 10 log